nimare/parcellate.py

```diff
@@ -0,0 +1,166 @@
+        elif not r and not n:
+            raise ValueError("Either 'r' or 'n' must be provided.")
+
+        self.meta_estimator = meta_estimator
+        self.target_image = target_image
+        self.n_clusters = listify(n_clusters)
+        self.r = r
+        self.n = n
+
+    def _preprocess_input(self, dataset):
+        """Mask required input images using either the dataset's mask or the estimator's.
+
+        Also, insert required metadata into coordinates DataFrame.
+        """
+        super()._preprocess_input(dataset)
+
+        # All extra (non-ijk) parameters for a kernel should be overrideable as
+        # parameters to __init__, so we can access them with get_params()
+        kt_args = list(self.kernel_transformer.get_params().keys())
+
+        # Integrate "sample_size" from metadata into DataFrame so that
+        # kernel_transformer can access it.
+        if "sample_size" in kt_args:
+            self.inputs_["coordinates"] = add_metadata_to_dataframe(
+                dataset,
+                self.inputs_["coordinates"],
+                metadata_field="sample_sizes",
+                target_column="sample_size",
+                filter_func=np.mean,
+            )
+
+    @use_memmap(LGR, n_files=1)
+    def _fit(self, dataset):
+        """Perform coordinate-based coactivation-based parcellation on dataset.
+
+        Parameters
+        ----------
+        dataset : :obj:`nimare.dataset.Dataset`
+            Dataset to analyze.
+        """
+        self.dataset = dataset
+        self.masker = self.masker or dataset.masker
+        self.null_distributions_ = {}
+
+        # Loop through voxels in target_mask, selecting studies for each and running MACMs (no MCC)
+        target_ijk = np.vstack(np.where(self.target_mask.get_fdata()))
+        target_xyz = vox2mm(target_ijk, self.masker.mask_img.affine)
+        for i_coord in target_xyz.shape[1]:
+            xyz = target_xyz[:, i_coord]
+            macm_ids = dataset.get_studies_by_coordinate(xyz, r=self.r, n=self.n)
+            coord_dset = dataset.slice(macm_ids)
+
+            # This seems like a somewhat inelegant solution
+            # Check if the meta method is a pairwise estimator
+            if "dataset2" in inspect.getfullargspec(self.meta_estimator.fit).args:
+                unselected_ids = sorted(list(set(dataset.ids) - set(macm_ids)))
+                unselected_dset = dataset.slice(unselected_ids)
+                self.meta_estimator.fit(coord_dset, unselected_dset)
+            else:
+                self.meta_estimator.fit(coord_dset)
+            macm_data = self.meta_estimator.results.get_map(self.target_image, return_type="array")
+            if i_coord == 0:
+                data = np.zeros((target_xyz.shape[1], len(macm_data)))
+            data[i_coord, :] = macm_data
+
+        # Correlate voxel-wise MACM results
+        data = np.corrcoef(data)
+
+        # Convert voxel-wise correlation matrix to distances
+        data = 1 - data
```

Comment:
I'm not sure if [sklearn.cluster.KMeans](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html#sklearn.cluster.KMeans) automatically calculates the distance values or not. There's a `precompute_distances` parameter, but it's deprecated and has no effect.

Reply:
After looking through sklearn's examples a bit, I think it calculates distance internally, so we can feed it the raw ALE values.